stm32/libcanard/bxcan/src/bxcan.c

```diff
@@ -0,0 +1,753 @@
+    volatile BxCANType* bxcan_base  = NULL;
+    bool*               error_iface = NULL;
+    if (iface_index == 0U)
+    {
+        bxcan_base  = BXCAN1;
+        error_iface = &g_error_iface0;
+    }
+    else if ((iface_index == 1U) && (BXCAN_MAX_IFACE_INDEX == 1U))
+    {
+        bxcan_base  = BXCAN2;
+        error_iface = &g_error_iface1;
+    }
+    else
+    {
+        input_ok = false;  // Invalid CAN interface number.
+    }
+
+    // Registers any error states for the selected CAN interface. It also handles the requirement
+    // to automatically abort all pending transmissions if the CAN controller enters the bus-off state.
+    if (input_ok == true)
+    {
+        processErrorStatus(bxcan_base, error_iface);
+    }
+
+    // Validate the other inputs.
+    if (payload_size > 8U)
+    {
+        input_ok = false;  // Payload size for classic CAN must be <= 8 bytes.
+    }
+
+    if ((payload_size > 0U) && (payload == NULL))
+    {
+        input_ok = false;  // NULL pointer payload with non-zero payload size.
+    }
+
+    if ((extended_can_id & (uint32_t)(!(BXCAN_FRAME_EXT_ID_MASK))) != 0U)
+    {
+        input_ok = false;  // Extended_can_id must be exactly 29 bits, thus 3 MSB of 32-bit word must be 0.
+    }
+
+    // When current_time > deadline, the frame is discarded, TX timeout error is registered, and true is returned.
+    if ((input_ok == true) && (current_time > deadline))
+    {
+        *error_iface = true;   // TX timeout error occurred.
+        input_ok     = false;  // Skip the further data transmission steps.
+        out_status   = true;   // We must return true when discarding the frame. It is considered sent.
+    }
+
+    // Seeking an empty slot, checking if priority inversion would occur if we enqueued now.
+    // We can always enqueue safely if all TX mailboxes are free and no transmissions are pending.
+    uint8_t tx_mailbox = 0xFF;  // 0xFF indicates no free mailboxes or a priority inversion.
+    if (input_ok == true)
+    {
+        static const uint32_t AllTME   = BXCAN_TSR_TME0 | BXCAN_TSR_TME1 | BXCAN_TSR_TME2;
+        static const uint32_t TIMEMASK = BXCAN_TDTR_TIME_MASK >> BXCAN_TDTR_TIME_SHIFT;
+
+        if ((bxcan_base->TSR & AllTME) != AllTME)  // At least one TX mailbox is used, detailed check is needed
+        {
+            const bool tme[3U] = {(bxcan_base->TSR & BXCAN_TSR_TME0) != 0U,
+                                  (bxcan_base->TSR & BXCAN_TSR_TME1) != 0U,
+                                  (bxcan_base->TSR & BXCAN_TSR_TME2) != 0U};
+
+            for (uint8_t i = 0U; i < 3U; i++)
+            {
+                // Abort the selected mailbox if it is pending and expired. The timeout error is reported.
+                // This frees-up stale slots. The time-out value in the TDTR register is 16 bits.
+                if ((current_time & TIMEMASK) >
+                    ((bxcan_base->TxMailbox[i].TDTR & BXCAN_TDTR_TIME_MASK) >> BXCAN_TDTR_TIME_SHIFT))
```

Comment:
This thing with the time register is not going to work.

1. You seem to be assuming here `(current_time & TIMEMASK)` that the time units of the API match those of the time register. This is not what the API docs advertise; they allow arbitrary time units (usually microseconds).

2. The time register is only 16 bits wide, meaning that it will be certain to overflow in very practical applications.

3. I could not find where you handle the timeout value.

The hardware timestamping feature is pretty useless in bxCAN; it's hard to go wrong by pretending that it doesn't exist, especially considering the fact that it is essentially declared useless in the errata.

What I suggest instead is to use the old libuavcan v0 approach where we simply keep a global array of timeout values per TX mailbox (so six elements overall, three per iface), and we write a new value in there whenever we begin transmission of a new frame. We check the timeout early in `bxCANPush()` and abort transmission if timed out; this may also free up a slot before the new frame is scheduled. This way is also logically much simpler than dealing with hardware timestamping.

Reply:
I totally missed this one. In the current code, I have changed it as you describe.